[Feature] Display of username and phone-number + [Refactoring] HTTP-requests, Json-mapping and error-handling

loewenzahm opened the pull request:
Desription: After a successfull login, personal information (i.e. username and phone-number) are displayed above the list for bikes.

Steps: Extract information from user's json-object; Store info in shared preferences; Enhance/Update view of the MainActivity

Edit: issue #16

cybre-finn committed:
initial commit

cybre-finn committed:
add login activity

cybre-finn committed:
change to dark material

cybre-finn committed:
tidy up

cybre-finn committed:
add RentActivity

cybre-finn committed:
direct Floating Button to RentActivity

cybre-finn committed:
change FloatingButton drawable

cybre-finn committed:
RentActivity improve navigation

cybre-finn committed:
add rental form

cybre-finn committed:
specify gitignore

cybre-finn committed:
adjust colors

cybre-finn committed:
fix missing margins

cybre-finn committed:
abstract AsyncTask

cybre-finn committed:
tidy up

cybre-finn committed:
enhance abstraction of RequestHandler

cybre-finn committed:
modify gitignore

cybre-finn committed:
implement list of rented bikes

cybre-finn committed:
adapt LoginActivity to changes

cybre-finn committed:
string name: "response" better than "result"

cybre-finn committed:
give RentActivity real functionality

cybre-finn committed:
add return activity

cybre-finn committed:
add readme

cybre-finn committed:
fix listing

cybre-finn committed:
Create LICENSE

cybre-finn committed:
edit README

cybre-finn committed:
update gitignore

cybre-finn committed:
update README

cybre-finn committed:
bump version: 0.1.1

cybre-finn committed:
add icon

cybre-finn committed:
bump version: 0.1.2

cybre-finn committed:
add Izzy's description to README

cybre-finn committed:
Remove validation of pin length. Closes #6

Poussinou committed:
Update README.md

cybre-finn committed:
Merge pull request #8 from Poussinou/patch-1

Update README.md

cybre-finn committed:
tidy up: remove unneeded dependencies, random debug logs etc...

cybre-finn committed:
modify README: Privacy etc...

cybre-finn committed:
README: point out that it's unofficial

cybre-finn committed:
remove logical not from faulty template

cybre-finn committed:
replace mocky settings with working logout

cybre-finn committed:
bump version: 0.1.3

cybre-finn committed:
fix the missed "action_logout"

cybre-finn committed:
add map option, fixes #11

cybre-finn committed:
add custom string to action_map

cybre-finn committed:
add indicator message for no bikes

cybre-finn committed:
bump version: 0.1.4

cybre-finn committed:
bump version: 0.1.5

cybre-finn committed:
list Kassel as supported (#4)

cybre-finn committed:
list Poznań as supported

cybre-finn committed:
various changes to README

cybre-finn committed:
grade version update

cybre-finn committed:
update README

cybre-finn committed:
add support for electric lock && gps tracked

cybre-finn committed:
fix no gps_tracking data available

cybre-finn committed:
add city to README

cybre-finn committed:
bump version: 0.1.6

cybre-finn committed:
add some cities as proposed in #4

Commit:
Adds intent listener for nxtb.it links

Commit:
Adds app name to open with dialog

cybre-finn committed:
remove return bike comment, fixes #21

Commit:
Bumps version to 0.1.7.1

Commit:
Removes comments

Commit:
Bumps to version 0.1.7.2

loewenzahm committed:
[Bugfix] Fix logout bug

On a logout the login-key has not been removed from the shared preferences due to a typo ("loginkey" instead of "loginKey"). Now that the typo is fixed the login key is no longer kept stored after logouts which resolves the bug mentioned in issue #16

cybre-finn committed:
Merge pull request #31 from loewenzahm/master

[Bugfix] Fix logout bug

loewenzahm committed:
[Feature] Add displaying of username and phone-number after login [issue #16]

Desription: After a successfull login, personal information (i.e. username and phone-number) are displayed above the list for bikes.

Steps: Extract information from user's json-object; Store info in shared preferences; Enhance/Update view of the MainActivity

h0chi commented:
Neat! Unfortunately, this is a larger chunk of code that cannot be reviewed so easily. So I'll leave this here for later, when I can try this on an Android phone. (or @lgehr could do this as well)
Feel free to push me if I'm taking to long. I tend to procrastinate reviews in this repo. ^^

loewenzahm commented:
By the way, the feature requires not only to be logged in but to log in. That means that for already logged in users, the MainActivity will show "unknown" as name and phone number. But that changes immediately after they log out and in again.

That is because the personal information are only requested in the LoginActivity and then stored in the SharedPreferences. The MainActivity then only uses the values stored in the SharedPreferences.

In a future version we could add a function in the MainActivity that checks if the loginKey is stored in the SharedPreferences for situations in which the personal data is missing. In this case the personal information could be requested via the api in the MainActivity as well. That would solve the situation for already logged in users (i.e. users that have a loginKey but no personal info in the SharedPreferences).

loewenzahm committed:
[Refactoring] Use Volley for HTTP-request and Jackson for Json-mapping in the LoginActivity

This is done for the LoginActivity exemplary - if the review shows that this approach fits for all programmers we could use this for all requests in all other Activities as well


parts of issue #19 (Volley + Jackson)

parts of issue #13 (proper error handling using snackbar-messages)

loewenzahm commented:
@h0chi & @lgehr : 
I refactored the LoginActivity (HTTP-requests with Volley + Json-mapping with Jackson + Error displaying using Snackbar). We/I could do that for the other activities as well if you are okay with the way it is done now. Feel free to ask any questions or if you would like to talk about the changes.

lgehr commented:
Thanks for the PR. 🎉🙏
I did not have time to test everything in detail but I have some notes:

For [50de8c7](https://github.com/h0chi/next-companion/pull/32/commits/50de8c7114adb806995fa51fd05ed964d948044b):
I do like the feature. Although I don't really like that the save the userinformation locally. I would rather have it be requested every time. (Since the scree_name can change?!, I did not know that nextbike has usernames...).
I think also this feature should tell the user the information the server has of them rather then what the client assumes the server has.  
This would also fix the "bug" that the user has to logout and in to get the information.

For [3c7fa68](https://github.com/h0chi/next-companion/pull/32/commits/3c7fa6819af91be5aa5add5f2e6484254a317954):
Volley seems to be a great choice to implement new features quickly. 
For error handling we could send `show_errors=true` with our requests. Then the server will reply with an error-code which volley can use (Although I do not know what volley makes of it). In the future we could also display the error-msg from the server. See [nextbike-api](https://api.nextbike.net/api/documentation#error_codes)  
If we use a JSON serializing library I think it would make sense to also serialize our request and not only the replies with it. 
But only de-serializing the replies is maybe better then nothing?! :D


These are just my 2 cents. I think they should not stand in the way of merging something because I did not find the time to do any great contribution to this project.

Kind Regards!

loewenzahm commented:
I really appreciate your feedback and ideas - thanks! :pray: 

And I totally agree on your suggestion to access the user's data not only on login but every time the app is used - can't understand myself why I implemented it the other way. I will fix that in the next days. :)

And yes, further error handling should definitely be something to be considered in the future... :smile:

loewenzahm commented:
@lgehr    I tried to change the way (or better the situation in which) we access the user's name and phone number as discussed above. Unfortunately we can't access this information with another API call than the one we use to authenticate the user. The reason for that is that the API calls for getting user information seem not to work with our API key. I tried getting user information with both, the official API's method (getUserInfo) as well as with the endpoint listed in h0chi's reverse engineered API (api/v1.1/getUserDetails.json). For both approaches I get `error code="17" message="access denied for API key"`. Is it possible that our API key has limited access concerning that data, @h0chi ? (Other API calls work as usual, by the way.)

For now I would suggest that we keep it that way (despite you have an idea to solve that problem) and maybe merge the PR? :)

Commit:
Recommiting strings.xml files without old apikey

lgehr commented:
Hey. Sorry I had to close the PR to  force push a new commit which rewrote the git history.
Reason is that the nextbike people asked us to not have the apikey we used before in our history.

lgehr commented:
I think we get the merge errors now because of the forced history rewrite... 
@loewenzahm Could you look into this?

lgehr commented:
Hey @meierjan 
could we get access to `getUserInfo` with our apikey? Thanks!

meierjan commented:
Done. Might take an hour for the cache for invalidate. :) On Mar 20, 2022 11:42 PM, lgehr ***@***.***> wrote:
Hey @meierjan
could we get access to getUserInfo with our apikey? Thanks!

—Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.
You are receiving this because you were mentioned.Message ID: ***@***.***>

loewenzahm committed:
Merge branch 'master' into master

loewenzahm commented:
> I think we get the merge errors now because of the forced history rewrite... @loewenzahm Could you look into this?

Done. :)

Commit:
Try to retrieve user information (name + phoneNumber) from within the MainActivity everytime the Activity starts instead of doing this only once on login

loewenzahm commented:
> For [50de8c7](https://github.com/h0chi/next-companion/pull/32/commits/50de8c7114adb806995fa51fd05ed964d948044b): I do like the feature. Although I don't really like that the save the userinformation locally. I would rather have it be requested every time. (Since the scree_name can change?!, I did not know that nextbike has usernames...). I think also this feature should tell the user the information the server has of them rather then what the client assumes the server has. This would also fix the "bug" that the user has to logout and in to get the information.
>

@lgehr  I addressed that concern with 881f95d2b9665c3ba7c51d11d69b231fb6cb3899 . Now these information are requested from within the MainActivity every time they are needed instead of being stored locally. Thanks for setting the base for this solution by asking for the enhancement of our api key. :smile: 

@meierjan Thank you as well for giving us access to the relevant data  :pray:

loewenzahm commented:
@lgehr @h0chi   Do you think you could merge this or are there other things I must change first? :)

lgehr commented:
Sorry for the really late response. 

I have problem applying the patch to my local repo. I also find it rather strange that this PR has 69 commits, because of the history rewrite. 
I would like for this PR to be just the one commit is really provides. (It hopefully will be easier to apply it locally then.)

Maybe you could make a new PR from a new branch which just cherry-picks the topcommit. Thanks!

h0chi commented:
Any idea what happened here? B/c at least from the github GUI the commits (the two I checked) including the initial have the same hashes...

h0chi commented:
I looked into it and it looks like @loewenzahm somehow managed to have a copy of every commit (as in reapplied?)  in their git history. This breaks the commit graph of @loewenzahm's repo and should not be committed into our master.

Possible solutions are: 
- @loewenzahm  could try to somehow cherry-pick the top commit which contains the relevant feature as @lgehr said.
- We could try to squash and merge this PR (however, I would prefer to do this locally or to another branch)
- @loewenzahm could delete their repo, close this PR, fork again and manually apply and commit their change again, then open another PR.

I could fix and merge this locally, but I do not know if the feature is mature.
